src/hotspot/share/gc/shenandoah/shenandoahHeapRegion.cpp

```diff
@@ -436,16 +501,29 @@ ShenandoahHeapRegion* ShenandoahHeapRegion::humongous_start_region() const {
 }
 
 void ShenandoahHeapRegion::recycle() {
+  ShenandoahHeap* heap = ShenandoahHeap::heap();
+
+  if (affiliation() == YOUNG_GENERATION) {
+    heap->young_generation()->decrease_used(used());
+  } else if (affiliation() == OLD_GENERATION) {
+    heap->old_generation()->decrease_used(used());
+  }
+
   set_top(bottom());
   clear_live_data();
 
   reset_alloc_metadata();
 
-  ShenandoahHeap::heap()->marking_context()->reset_top_at_mark_start(this);
+  heap->marking_context()->reset_top_at_mark_start(this);
   set_update_watermark(bottom());
 
   make_empty();
-  _affiliation = ShenandoahRegionAffiliation::FREE;
+#ifdef DEBUG_TRACE
+  printf("SHR::recycle(), setting region (%llx, %llx, %llx) to FREE\n",
+         (unsigned long long) bottom(), (unsigned long long) top(), (unsigned long long) end());
+  fflush(stdout);
```

Comment:
In Hotspot, it's simpler to use tty->print_cr("...") for this, or even better use logging e.g. log_trace(gc)("...") or so.

Reply:
I agree, I've taken out all the debug scaffolding.